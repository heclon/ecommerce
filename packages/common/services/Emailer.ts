import Email from 'email-templates'
import { createTransport, Transporter } from 'nodemailer'
import { minify } from 'html-minifier'

export default class Emailer {
  private readonly transport: Transporter
  constructor() {
    const { MAIL_SERVICE, MAIL_AUTH_USER, MAIL_AUTH_PASSWORD } = process.env
    if (!MAIL_SERVICE || !MAIL_AUTH_USER || !MAIL_AUTH_PASSWORD)
      throw new Error('Unable to create transport for email service as credentials were not provided')
    const options = {
      service: MAIL_SERVICE,
      auth: {
        user: MAIL_AUTH_USER,
        pass: MAIL_AUTH_PASSWORD,
      },
    }
    this.transport = createTransport(options)
    this.transport.use('compile', async (mail, callback) => {
      mail.data.html = await minify(mail.data.html as string, {
        collapseBooleanAttributes: true,
        collapseInlineTagWhitespace: true,
        collapseWhitespace: true,
        minifyCSS: true,
        removeAttributeQuotes: true,
        removeTagWhitespace: true,
      })
      callback()
    })
  }

  private newEmail() {
    return new Email({
      message: {
        from: 'no-reply@providerchoice.com.au',
      },
      // Set the 'send' property to true to send emails in develop and test environments
      // send: true,
      preview: false,
      views: {
        options: {
          extension: 'lodash',
        },
      },
      transport: this.transport,
    })
  }

  async sendServiceRequestInterest(
    requestId: string,
    requestCreatedAt: string,
    supportName: string,
    participantFirstName: string,
    providerEmail: string
  ) {
    return this.newEmail().send({
      template: 'service-request-interest',
      message: {
        to: 'providers@ecommerce.com.au',
      },
      locals: {
        requestId,
        requestCreatedAt,
        supportName,
        participantFirstName,
        providerEmail,
      },
    })
  }
}
