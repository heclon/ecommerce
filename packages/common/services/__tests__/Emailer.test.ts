import Emailer from '../Emailer'

describe('Emailer', () => {
  test('sendServiceRequestInterest', async () => {
    const emailer = new Emailer()
    const response = await emailer.sendServiceRequestInterest(
      'fake-request-is',
      '11/11/2022',
      'Test Support',
      'Steve',
      'test@provider.com.au'
    )

    expect(response.originalMessage).toMatchSnapshot()
  })
})
