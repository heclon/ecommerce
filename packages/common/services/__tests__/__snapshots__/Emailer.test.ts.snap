// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emailer sendServiceRequestInterest 1`] = `
Object {
  "attachments": Array [],
  "from": "no-reply@providerchoice.com.au",
  "html": "    <div style=\\"border: thin solid lightgrey; font-family: Arial, Helvetica, sans-serif;\\">
      <div style=\\"padding:1rem; background-color: #4E42D3; color: white;\\">
        <h2>Service Request Interest</h2>
      </div>
      <div style=\\"padding: 1rem;\\">
        <p>The following provider has shown an interest in a service request.</p>
        <p><span style=\\"font-weight: bold;\\">Provider email:</span> test@provider.com.au</p>
        <p><span style=\\"font-weight: bold;\\">Request ID:</span> fake-request-is</p>
        <p><span style=\\"font-weight: bold;\\">Request created:</span> 11/11/2022</p>
        <p><span style=\\"font-weight: bold;\\">Support Name:</span> Test Support</p>
        <p><span style=\\"font-weight: bold;\\">Participant given name:</span> Steve</p>
      </div>
    </div>",
  "subject": "Service Request Interest: Steve - Test Support (fake-request-is)",
  "text": "SERVICE REQUEST INTEREST

The following provider has shown an interest in a service request.

Provider email: test@provider.com.au

Request ID: fake-request-is

Request created: 11/11/2022

Support Name: Test Support

Participant given name: Steve",
  "to": "providers@ecommerce.com.au",
}
`;
